pkg/kubelet/kubelet.go

```diff
@@ -2037,6 +2032,7 @@ func (kl *Kubelet) HandlePodRemoves(pods []*v1.Pod) {
 	start := kl.clock.Now()
 	for _, pod := range pods {
 		kl.podManager.DeletePod(pod)
+		kl.podKiller.MarkMirrorPodPendingTermination(pod)
```

Comment:
In the next update, I would do the following:

1. use slice instead of channel to track the outstanding pod killing requests
2. func names in pod killer no longer carry 'mirror'
3. podTerminationMap and the killing Set are combined into one map
4. remove the above call in HandlePodRemoves
5. remove Close() func from PodKiller

Reply:
if you don't use a channel, you'll have to have some sort of signalling method to wake up the podKiller goroutine.  what did you have in mind? polling the queue?

Reply:
The podKiller goroutine would continue to completion once launched.

I think you are asking how the subsequent pod killing requests are retrieved.
podKillerWithSlice#PerformPodKillingWork() would copy outstanding pod killing requests from the slice field.
If there is no pod killing request at the moment, PerformPodKillingWork() would return.

It would be periodically run:
```
	go wait.Until(kl.podKiller.PerformPodKillingWork, 1*time.Second, wait.NeverStop)
```

Reply:
that would introduce up to 1s of latency between a call to `KillPod()` and the podkiller actually starting the goroutine to kill it.  where a channel is more or less instantaneous.

Reply:
I switched to using slice because channel doesn't provide 'peek' functionality: we have to fetch the request out of the channel.
For podKillerWithSlice, once pod killing request is sent, we know where to find pod pending termination.

If incurring lower latency is of consideration, we can shorten the period in wait.Until() to, e.g. 500 milliseconds.

I am open to other suggestion on this as well.

Reply:
we don't want to poll at any rate when the previous behavior was ~instant.  just keep the channel to avoid regressions.  we want to change _as little as possible_ in order to fix the issue.  strapping a map with redundant data onto the side of the existing data path is preferable to changing the channel based code.

Reply:
Looking at podKillerWithChannel#MarkMirrorPodPendingTermination(), it has check which limits the addition to pod which has mirror pod:
```
		if _, ok := pk.getMirrorPodByPod(pod); ok {
```
So the entries in mirrorPodTerminationMap would be for mirror pod. Hence I name the map that way.

Reply:
@sjenning 
Let me know if you have more review comments.

Thanks